app/javascript/dashboard/helper/editorHelper.js

```diff
@@ -281,3 +283,109 @@ export function setURLWithQueryAndSize(selectedImageNode, size, editorView) {
+ * @returns {Object} - The created mention node and the updated from and to positions
+ */
+const getMentionNode = (editorView, content, from, to) => {
+  const node = editorView.state.schema.nodes.mention.create({
+    userId: content.id,
+    userFullName: content.name,
+  });
+  return { node, from, to };
+};
+
+/**
+ * Creates a canned response node for the editor state
+ * @param {Object} editorView - The editor view instance
+ * @param {string} content - The canned response content
+ * @param {number} from - The start position of the canned response in the document
+ * @param {number} to - The end position of the canned response in the document
+ * @param {Object} variables - The variables to replace in the canned response
+ * @returns {Object} - The created canned response node and the updated from and to positions
+ */
+const getCannedResponseNode = (editorView, content, from, to, variables) => {
+  const updatedMessage = replaceVariablesInMessage({
+    message: content,
+    variables,
+  });
+
+  const node = new MessageMarkdownTransformer(messageSchema).parse(
+    updatedMessage
+  );
+
+  from = node.textContent === updatedMessage ? from : from - 1;
+
+  return { node, from, to };
+};
+
+/**
+ * Creates a variable node for the editor state
+ * @param {Object} editorView - The editor view instance
+ * @param {string} content - The variable content
+ * @param {number} from - The start position of the variable in the document
+ * @param {number} to - The end position of the variable in the document
+ * @returns {Object} - The created variable node and the updated from and to positions
+ */
+const getVariableNode = (editorView, content, from, to) => {
+  const node = editorView.state.schema.text(`{{${content}}}`);
+  return { node, from, to };
+};
+
+/**
+ * Creates an emoji node for the editor state
+ * @param {Object} editorView - The editor view instance
+ * @param {string} emoji - The emoji content
+ * @param {number} from - The start position of the emoji in the document
+ * @param {number} to - The end position of the emoji in the document
+ * @returns {Object} - The created emoji node and the updated from and to positions
+ */
+const getEmojiNode = (editorView, emoji, from, to) => {
+  const node = editorView.state.schema.text(emoji);
+  return { node, from, to };
+};
+
+/**
+ * Creates a content node based on the type (mention, canned response, variable, or emoji)
+ * @param {Object} editorView - The editor view instance
+ * @param {string} type - The type of content (mention, cannedResponse, variable, or emoji)
+ * @param {string|Object} content - The content to create the node from
+ * @param {Object} range - The range object containing the from and to positions
+ * @param {Object} variables - The variables to replace in the canned response (optional)
+ * @returns {Object} - The created content node and the updated from and to positions
+ */
+export const getContentNode = (editorView, type, content, range, variables) => {
```

Comment:
I believe there's some repetition in the node creation functions that we can refactor. Also the `getContentNode` function can be simplified using a more functional approach

```
const createNode = (editorView, nodeType, content) => {
  const { state } = editorView;
  switch (nodeType) {
    case 'mention':
      return state.schema.nodes.mention.create({
        userId: content.id,
        userFullName: content.name,
      });
    case 'cannedResponse':
      return new MessageMarkdownTransformer(messageSchema).parse(content);
    case 'variable':
      return state.schema.text(`{{${content}}}`);
    case 'emoji':
      return state.schema.text(content);
    default:
      return null;
  }
};

const nodeCreators = {
  mention: (editorView, content, from, to) => ({ node: createNode(editorView, 'mention', content), from, to }),
  cannedResponse: (editorView, content, from, to, variables) => {
    const updatedMessage = replaceVariablesInMessage({ message: content, variables });
    const node = createNode(editorView, 'cannedResponse', updatedMessage);
    return { node, from: node.textContent === updatedMessage ? from : from - 1, to };
  },
  variable: (editorView, content, from, to) => ({ node: createNode(editorView, 'variable', content), from, to }),
  emoji: (editorView, content, from, to) => ({ node: createNode(editorView, 'emoji', content), from, to }),
};

export const getContentNode = (editorView, type, content, { from, to }, variables) => {
  const creator = nodeCreators[type];
  return creator ? creator(editorView, content, from, to, variables) : { node: null, from, to };
};

```

Note: Please test it properly updating the code.

Reply:
@muhsin-k Done https://github.com/chatwoot/chatwoot/pull/10055/commits/109a1400a6362acbd6437b3c0e3b9abffb339e49